libbpf-rs/src/probe.rs

```diff
@@ -0,0 +1,57 @@
+//! Probe for supported system features
+//!
+//! Note that running many of these probes in a short amount of time can cause
+//! the kernel to reach the maximal size of lockable memory allowed for the
+//! user, causing subsequent probes to fail. In this case, the caller may want
+//! to adjust that limit with setrlimit().
+
+use super::map::MapType;
+use super::program::ProgramType;
+use libbpf_sys::bpf_func_id;
+
+use crate::*;
+
+/// Detects if host kernel supports BPF programs of a given type
+/// * `prog_type` - BPF program type to detect kernel support for
+///
+/// Make sure the process has required set of CAP_* permissions (or runs as
+/// root) when performing feature checking.
+pub fn probe_prog_type(prog_type: ProgramType) -> Result<bool> {
+    let ret = unsafe { libbpf_sys::libbpf_probe_bpf_prog_type(prog_type as u32, std::ptr::null()) };
+    match ret {
+        0 => Ok(false),
+        1 => Ok(true),
+        _ => Err(Error::System(-ret)),
+    }
+}
```

Comment:
I can see why you'd arrange functionality the way you did, but perhaps an alternative design would be to extend `ProgramType` with a method to query whether it is supported (similar for other types)? Then users would not need to import this functionality explicitly, it's just generally easier discoverable, and it would allow for much more explicit naming (e.g., `impl ProgramType { fn is_supported(&self) -> Result<bool> { ... } }` or `fn is_supported_on_host(&self)`; in addition to `is_helper_supported(&self, helper_id...) ...`).

I am not a fan of the probe terminology as the sole indication of what a function is doing, as it is ambiguous. Again, I can see where it's coming from, but I don't think we should sacrifice explicitness & proper integration (native feel) with the binding language, just to preserve design decision made in a different language and that users may have no awareness of.

How does that sound to you?

Reply:
good idea - i'll give it a try